### PR TITLE
 Binary File Storage Research

### PR DESCRIPTION
**Pull request by adamfilipow92**
Contains investigation results of #849 

Full documentation can be found in the file: host_binaries_solutions_comparison.md 

# Host binaries solutions comparison

## Conclusion

By comparing these three solutions, talk with @jan-gogo, and some amount of test git-annex we staying with GitHub releases and automated the release process.

## Checklist

- [X] Documented git-lfs
- [X] Documented git-submodules
- [X] Documented git-annex
- [X] Comparison of pros and cons
- [X] Check how to configure git-annex with google drive
- [x] Create a test repository and try to sync with another teammate

### CODE REVIEW

**Comment by bootstraponline** [comment]
LFS bandwidth is expensive and limited https://docs.github.com/en/github/managing-large-files/about-storage-and-bandwidth-usage

I'm not sure we'll be able to use github LFS, maybe we could consider setting up our own lfs server?

**Comment by adamfilipow92** [comment]
> GitHub LFS bandwidth is expensive and limited https://docs.github.com/en/github/managing-large-files/about-storage-and-bandwidth-usage
> 
> I'm not sure we'll be able to use github LFS, maybe we could consider setting up our own lfs server?

> https://docs.github.com/en/github/managing-large-files/about-storage-and-bandwidth-usage

Thanks for link I added it to my docs. 

I found how to configure own lfs server [here](https://docs.github.com/en/enterprise/2.17/admin/installation/configuring-git-large-file-storage-to-use-a-third-party-server) but its about enterprise server and i cannot found any options for other plans. I contact with github and I ask about it. 

Git-annex can store files on different storages ftp, amason s3 and other. [Here](https://git-annex.branchable.com/special_remotes/ ) is list of supported storage types. But first I need to check how configure it and how it working.

**Comment by bootstraponline** [comment]
> I found how to configure own lfs server [here](https://docs.github.com/en/enterprise/2.17/admin/installation/configuring-git-large-file-storage-to-use-a-third-party-server) but its about enterprise server and i cannot found any options for other plans. I contact with github and I ask about it.

I know GitLab setup their own open source LFS server.
https://about.gitlab.com/blog/2015/11/23/announcing-git-lfs-support-in-gitlab/

**Comment by bootstraponline** [comment]
git submodules are free in terms of dollar cost. They are super painful to work with though as a developer. 🙂

**Comment by bootstraponline** [comment]
https://github.com/git-lfs/git-lfs/wiki/Implementations has lfs implementations

**Comment by bootstraponline** [comment]
Ideally Flank is able to host binaries using a free hosted option. One idea is to treat each binary as an artifact that we publish to GitHub packages:
https://github.com/features/packages

that'd provide unlimited free bandwidth for open source repos.

**Comment by adamfilipow92** [comment]
> https://github.com/git-lfs/git-lfs/wiki/Implementations has lfs implementations

Yes, GitHub support send me this link on friday when I was after work. Today I update description.


> Ideally Flank is able to host binaries using a free hosted option. One idea is to treat each binary as an artifact that we publish to GitHub packages:
> https://github.com/features/packages
> 
> that'd provide unlimited free bandwidth for open source repos.

We try git-annex with google drive it's should be free and enough for our purposes.

**Comment by adamfilipow92** [comment]
We try to test google drive with @jan-gogo but only shared drive can work with the many google accounts. If I share directory on my google drive I can not download files from another account even If a directory is public for all. But there is a problem when people outside of the team want access to shared drive they need to ask about permissions, we need to accept it and it can take some time so I think this not a good solution.

I ask on rclone slack how to use a public directory (maybe they have any potential solution).

We try to find a free ftp server that can be configured to anonymous read-only access or where we can configure a read-only account but with no effects. 

For now, we pause research and wait for a response from rclone slack.

**Comment by bootstraponline** [comment]
https://docs.github.com/en/github/managing-large-files/distributing-large-binaries

**Comment by digitalbuddha** [comment]
Hi folks. We use flank at dropbox. Maybe y'all want to store your artifacts within dropbox? We have sdks in most languages and allow publicly sharing a file/folder without the receiver needing an account. https://help.dropbox.com/files-folders/share/share-outside-dropbox

Default space is 2gig but I can permanently donate more if y'all hit the limit.

**Comment by adamfilipow92** [comment]
> Hi folks. We use flank at dropbox. Maybe y'all want to store your artifacts within dropbox? We have sdks in most languages and allow publicly sharing a file/folder without the receiver needing an account. https://help.dropbox.com/files-folders/share/share-outside-dropbox
> 
> Default space is 2gig but I can permanently donate more if y'all hit the limit.

Hello! Thanks for the support flank. I will check how to integrate git-annex with dropbox and I back to you.

**Comment by adamfilipow92** [comment]
@digitalbuddha, I made some tests. Dropbox is a great solution but when we working with it internally. But when someone outside the internal team wants to use a shared artifact needs to copy it to own drive. It's can generate some problems when we update artifacts in this case user should remove the directory and copy it again. Maybe there is any way to see the public folder (read-only) as a directory on dropbox but I cannot find it?

**Comment by jan-gogo** [comment]
Research for remote storage takes too long. I think we should put our focus those 3 tickets:
https://github.com/Flank/flank/issues/898
https://github.com/Flank/flank/issues/899
Automation generation of test artifacts is an important problem and should be resolved before integration with new remote storage. 
Until we cannot find consensus about new remote storage, I guess we can still use GitHub releases. This issues
https://github.com/Flank/flank/issues/900
describes how to we can improve test artifacts management on Github.

File: docs/host_binaries_solutions_comparison.md [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+|:-----------------:|:-----------------------------:|:--------------:|:----------------------------:|
+|       Costs       |   5$/month per pack (50gb)    |   Cannot find clear answer            | Depend on storage provider   |
+| Versioning Files  |           Yes                 |   Yes          |  Yes                         |
+| Branch support    |           Yes                                             |   No, they are just a pointer to a particular commit of the submodules repository. It's can be harder to manage than git lfs and git-annex. We shouldn't store binary files in git because it's not designed for it. This is a purpose why git-annex and git-lfs exists. Some about it [here](https://robinwinslow.uk/2013/06/11/dont-ever-commit-binary-files-to-git/#:~:text=It%27s%20important%20to%20never%20commit,size%20will%20still%20be%20large.) |  Yes |
+| Flexibility       |  Yes, you can set your lfs server. [Check here](https://github.com/git-lfs/git-lfs/wiki/Implementations)   |   No | Yes, can use different storage providers. |
+
+By comparing these three solutions and talk with @jan-gogo  in our opinion the best solution is git-annex because:
```
Comment (jan-gogo) [inline docs/host_binaries_solutions_comparison.md:14]: It's not actual. Currently, we are staying with GitHub releases, we need to automate release process.

File: docs/host_binaries_solutions_comparison.md [hunk 2]
```diff
@@ -0,0 +1,262 @@
+# Host binaries solutions comparison
+
+## tl;dr
+
+After research and test git-annex, we think it's not suitable for our purposes. We try to use GitHub Releases, documentation and automating process.
```
Comment (jan-gogo) [inline docs/host_binaries_solutions_comparison.md:5]: Conclusions / status shoud be on the bottom of document.

File: docs/host_binaries_solutions_comparison.md [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+
+git add app-debug.apk
+git commit -m "Add apk file"
+git push origin master
+
+```
+
+### How to play with branches
+
+Sub-modules are not on a branch. They are just a pointer to a particular commit of the sub-modules repository.
```
Comment (jan-gogo) [inline docs/host_binaries_solutions_comparison.md:130]: Actually there is an option to track specific branch on submodule.
https://stackoverflow.com/questions/9189575/git-submodule-tracking-latest

**Comment by bootstraponline** [comment]
I think we should move Flank to be a monorepo, and leverage GitHub releases / GitHub packages for binary management.

File: docs/host_binaries_solutions_comparison.md (hunk 3 above)
Comment (adamfilipow92) [reply docs/host_binaries_solutions_comparison.md]: Thanks, changed!

File: docs/host_binaries_solutions_comparison.md (hunk 2 above)
Comment (adamfilipow92) [reply docs/host_binaries_solutions_comparison.md]: Thanks, changed!

File: docs/host_binaries_solutions_comparison.md (hunk 1 above)
Comment (adamfilipow92) [reply docs/host_binaries_solutions_comparison.md]: Thanks, changed!

**Pull request merged**